### PR TITLE
Ran tests for Genres and added MaxLength.cs

### PR DESCRIPTION
**Pull request by brettmariani923**
Created and ran tests for Genres.sql, changed code to account for white spaces and empty strings. Created Max Length class within the LibraryManagement.Domain project to hold the max length value of a name insert.

### CODE REVIEW

File: LibraryManagement.SqlServerDatabase/DBO/Tables/Genres.sql [hunk 1]
```diff
@@ -2,4 +2,5 @@
 (
     GenreID INT CONSTRAINT PK_Genres PRIMARY KEY IDENTITY(1,1),
     Name VARCHAR(50) NOT NULL,
+    CONSTRAINT Name_NotEmpty CHECK (LEN(Name) > 0)
```
Comment (uhyeay2) [inline LibraryManagement.SqlServerDatabase/DBO/Tables/Genres.sql]: Lets change the name here, I looked up the naming convention for CHECK Constraints, I think we'd want something like: CHK_Genres_Name

https://www.scholarhat.com/tutorial/sqlserver/sql-server-naming-conventions-and-standards

File: LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+        [Fact]
+        public async Task InsertGenre_Given_GenreName_AlreadyExists_ShouldReturn_NoRowsUpdated()
+        {
+            var genreName = "Fiction";
+            await _dataAccess.ExecuteAsync(new InsertGenres(genreName));
+
+            var rowsAffected = await _dataAccess.ExecuteAsync(new InsertGenres(genreName));
+
+            Assert.Equal(1, rowsAffected);
```
Comment (uhyeay2) [inline LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs]: It looks like we are not asserting the correct thing here based on your test method name

File: LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        {
+            var request = new InsertGenres("");
+
+            await Assert.ThrowsAsync<SqlException>(async () => await _dataAccess.ExecuteAsync(request));
+        }
+
+        [Fact]
+        public async Task InsertGenre_Given_NameIsWhitespace_ShouldThrow_SqlException()
+        {
+            var request = new InsertGenres("   ");
```
Comment (uhyeay2) [inline LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs]: I love that you set up tests for NULL, Empty, and Whitespace. Great to make sure you know how to expect the application to behave in these scenarios. We can probably fit these all into one Test method though, using a 'Theory' instead of a 'Fact'.

It'd look something like this:
```
[Theory]
[InlineData(null!)]
[InlineData("")]
[InlineData("  ")]
public async Task InsertGenre_Given_NameIsInvalid_ShouldThrow_SqlException(string invalidName)
{
   var request = new InsertGenre(invalidName);
   
   await Assert.ThrowsAsync<SqlException>(async () => await _dataAccess.ExecuteAsync(request));
}

```

File: LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            await Assert.ThrowsAsync<SqlException>(async () => await _dataAccess.ExecuteAsync(request));
+        }
+
+        [Fact]
+        public async Task InsertGenre_WithTooLongName_ShouldThrowSqlException()
+        {
+            var longName = new string('A', 300);
```
Comment (uhyeay2) [inline LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs]: For this, instead of doing 300, lets change this to MaxLength.GenreName + 1 ? You might need another constraint on your Name column - or we could do the 'Max Length' validation/tests in the next layer up (Application Layer) when we get there. I'll leave it up to you on this one.

File: LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            var request = new InsertGenres(longName);
+
+            await Assert.ThrowsAsync<SqlException>(() => _dataAccess.ExecuteAsync(request));
+        }
+
+        [Fact]
+        public async Task InsertGenre_Given_NameWithSpecialCharacters_ShouldInsert_Successfully()
+        {
+            var specialChar = "Science & Technology!";
```
Comment (uhyeay2) [inline LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs]: the & is actually not really considered a 'special char' for sql - a special character would be something like Ç ü ⌂ é â ä å ç ê ë è ï î ì Ä Å É, etc.

These are only able to be inserted into the database if your column as an 'NVARCHAR' instead of 'VARCHAR' or 'CHAR'

File: LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+using LibraryManagement.Data.Tests.Helpers;
+using LibraryManagement.Data.Abstraction;
+using Microsoft.Data.SqlClient;
+using System.ComponentModel.DataAnnotations;
+using LibraryManagement.Domain.Constants;
+
+
+namespace LibraryManagement.Data.Tests.DataRequestTests.ExampleTests
+{
+    public class GenresTest : DataTest
```
Comment (uhyeay2) [inline LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs]: It'd be great if we could have the tests delete the inserted genre after they run. If you check your database you likely have multiple records for each test if you've been re-running them multiple times.

This would require you to make the DeleteGenre class, if you want to do the tests for that one in this same PR or in a followup PR I'll leave up to you, but it'd be good to purge out that old data after each test

File: LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/ExampleFetchByIdTests.cs [hunk 7]
```diff
@@ -23,9 +24,23 @@ public async Task ExampleFetchById_Given_RecordExists_ShouldReturn_ExistingRecor
...
+
+
+            // This class is just for an example, but ideally you would have multiple tests.
+            // You want to always consider the different possibilities/outcomes for the method/class you're testing, and then
+            // ensure that you have tests to cover for all scenarios. For example you might have one test to
+            // assert you get the record when it exists, but you might have another test that you get null when no record exists.
+        }
 }
+
+//make classes for each component of the DataRequest you want to test.
```
Comment (uhyeay2) [inline LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/ExampleFetchByIdTests.cs]: I think we can remove your comments from line 43 - 46

File: LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs (hunk 2 above)
Comment (brettmariani923) [reply LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs]: Ah I see. So with this I thought inserting the genreName value the first time returned the 1, and because it only returned a 1 and not 2 that seemed correct to me. I see now that its returning the 1 for the rowsAffected value, not for the initial value being inserted, so that means its still being added. 

I updated the Genres.sql code with a UNIQUE(Name) constraint so this stops happening. After that I changed the second insert in the code above to check for a SqlException instead. So it should be good to go now I think.

Comment (brettmariani923) [reply LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs:38]: The test is still failing, but its returning this error message  

Message: 
   Microsoft.Data.SqlClient.SqlException : Violation of UNIQUE KEY constraint 'UQ_Genres_Name'. Cannot insert duplicate key in object 'dbo.Genres'. The duplicate key value is (Fiction).
   The statement has been terminated.
   
So I think that means the program is working as intended, but it still isn't passing. I think instead of throwing an exception it is terminating the statement instead. I suppose I could just delete the test, but I'm open to any input you have on how to go forward.

Comment (brettmariani923) [reply LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs:38]: Here is the test method in case you wanted to take a look 

 [Fact]
 public async Task InsertGenre_Given_GenreName_AlreadyExists_ShouldReturn_NoRowsUpdated()
 {
     var genreName = "Fiction";

     var firstInsert = await _dataAccess.ExecuteAsync(new InsertGenres(genreName));

     await Assert.ThrowsAsync<SqlException>(async () => await _dataAccess.ExecuteAsync(new InsertGenres(genreName)));
 }

Comment (brettmariani923) [reply LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs:38]: Here is the updated SQL statement as well from Genres.sql

```CREATE TABLE [dbo].[Genres]
(
    GenreID INT CONSTRAINT PK_Genres PRIMARY KEY IDENTITY(1,1),
    Name VARCHAR(50) NOT NULL,
    CONSTRAINT CHK_Genres_Name CHECK (LEN(Name) > 0), 
    CONSTRAINT UQ_Genres_Name UNIQUE(Name)
);
```

File: LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs (hunk 6 above)
Comment (brettmariani923) [reply LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs]: I'll just do it in the same PR for simplicity.

```using LibraryManagement.Data.Abstraction;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LibraryManagement.Data.DataRequestObjects.Genres
{
    public class DeleteGenres : IDataExecute
    {
        public string _DeleteName { get; set; }

        public DeleteGenres(string name)
        {
            _DeleteName = name;
        }

        public object? GetParameters()
        {
            return new { Name = _DeleteName };
        }

        public string GetSql()
        {
            return "DELETE FROM Genres WHERE Name = @Name;";
        }
   
    }
}

Comment (brettmariani923) [reply LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs]: I'm going to rename the "GenresTest.cs" class to "InsertGenresTest.cs" and create a new class "DeleteGenresTest.cs" for that testing, just so it doesn't get too disorganized within the same class.

File: LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs (hunk 5 above)
Comment (brettmariani923) [reply LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs]: Sounds good, thats good to know. I updated it to NVARCHAR just in case something comes up down the road we didn't see ahead of time, like a weird sub genre we might decide to include. Its probably irrelevant though.

File: LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs (hunk 6 above)
Comment (brettmariani923) [reply LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs]: This is the extent of it I have so far, I had to create a GetGenre class in the DataRequestObjects (added it to the folder that holds InsertGenres and DeleteGenres) so I could fetch the column info to check if the delete test worked. I dont see much else that could go wrong beyond these tests. If you have anything that comes to mind let me know, I'll add it. 

```using LibraryManagement.Data.DataRequestObjects.Genres;
using LibraryManagement.Domain.Constants;
using Microsoft.Data.SqlClient;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using LibraryManagement.Data.DataRequestObjects.Examples;
using LibraryManagement.Data.Tests.Helpers;
using LibraryManagement.Data.Abstraction;
using System.ComponentModel.DataAnnotations;
using Azure.Core;

namespace LibraryManagement.Data.Tests.DataRequestTests.ExampleTests
{
    public class DeleteGenresTest : DataTest
    {
       [Fact]
        public async Task DeleteGenre_GivenCorrect_ShouldDelete_Successfully()
        {
            var genreName = "Philosophy";
        
            var request = new InsertGenres(genreName);
        
            var request2 = new DeleteGenres(genreName);
        
            var philosophy = new GetGenreByName(genreName);
        
            var result = await _dataAccess.FetchAsync(philosophy);
        
            Assert.Null(result);
        }

        [Fact]
        public async Task DeleteGenre_GivenConcurrent_ShouldDelete_Successfully()
        {
            var task1 = _dataAccess.ExecuteAsync(new InsertGenres("Action"));
            var task2 = _dataAccess.ExecuteAsync(new InsertGenres("Drama"));

            await Task.WhenAll(task1, task2);

            var deleteTask1 = _dataAccess.ExecuteAsync(new DeleteGenres("Action"));
            var deleteTask2 = _dataAccess.ExecuteAsync(new DeleteGenres("Drama"));

            await Task.WhenAll(deleteTask1, deleteTask2);

            var action = new GetGenreByName("Action");
            var drama = new GetGenreByName("Drama");

            var result1 = await _dataAccess.FetchAsync(action);
            var result2 = await _dataAccess.FetchAsync(drama);

            Assert.Null(result1);
            Assert.Null(result2);
        }
    }
}

File: LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs (hunk 4 above)
Comment (brettmariani923) [reply LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs:67]: I adjusted it to MaxLength + 1 like you said, and it passed the test. I'm not sure what you mean for the the constraint on the name column, do you mean to limit the Max Length of characters?

File: LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs (hunk 6 above)
Comment (brettmariani923) [reply LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs]: I also added a DeleteGenres() method after each relevant instance in the InsertGenresTest class, so there shouldn't be an issue with left over inserts now. Shouldn't be a problem going forward.

File: LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs (hunk 4 above)
Comment (brettmariani923) [reply LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/GenresTest.cs]: I adjusted it to MaxLength + 1 like you said, and it passed the test. For this one lets wait until we're on the next layer up. Its just running more basic tests to make sure the max length class does what we want it to do, right? Doesn't seem like it'll be too much trouble if we wait

File: LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/ExampleFetchByIdTests.cs (hunk 7 above)
Comment (brettmariani923) [reply LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/ExampleFetchByIdTests.cs]: hey I forgot to delete this in my latest push. I'll get in the next one.

File: LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/InsertGenresTest.cs [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        [Fact]
+        public async Task InsertGenre_GivenCorrect_ShouldInsert_Successfully()
+        {
+            var request = new InsertGenres("Philosophy");
+
+            var result = await _dataAccess.ExecuteAsync(request);
+
+            Assert.True(result == 1);
+
+            var deleteTask1 = _dataAccess.ExecuteAsync(new DeleteGenres("Philosophy"));
```
Comment (uhyeay2) [inline LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/InsertGenresTest.cs]: you can use request.Name instead of Philosophy in case you change it on line 22 above you won't have to change it in both places

Also it looks like you are not awaiting the call to delete the genre - it'd be good to await the call to ensure execution completes.

File: LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/InsertGenresTest.cs [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        }
+
+        [Fact]
+        public async Task InsertGenre_Given_GenreName_AlreadyExists_ShouldReturn_NoRowsUpdated()
+        {
+            var genreName = "Fiction";
+
+            await _dataAccess.ExecuteAsync(new InsertGenres(genreName));
+
+            await Assert.ThrowsAsync<SqlException>(async () => await _dataAccess.ExecuteAsync(new InsertGenres(genreName)));
```
Comment (uhyeay2) [inline LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/InsertGenresTest.cs]: The assertion doesn't seem to match the name of the test method

File: LibraryManagement.Data/DataRequestObjects/Genres/DeleteGenres.cs [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+        }
+
+        public object? GetParameters()
+        {
+            return new { Name = _DeleteName };
+        }
+
+        public string GetSql()
+        {
+            return "DELETE FROM Genres WHERE Name = @Name;";
```
Comment (uhyeay2) [inline LibraryManagement.Data/DataRequestObjects/Genres/DeleteGenres.cs]: I think it'd be better for us to delete by Id... That way if we have to genres with the same name but we only want to delete one we can ensure only that one is deleted

File: LibraryManagement.Data/DataRequestObjects/Genres/GetGenre.cs [hunk 11]
```diff
@@ -0,0 +1,30 @@
+﻿using LibraryManagement.Data.Abstraction;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace LibraryManagement.Data.DataRequestObjects.Genres
+{
+    public class GetGenreByName : IDataFetch<string>
```
Comment (uhyeay2) [inline LibraryManagement.Data/DataRequestObjects/Genres/GetGenre.cs]: If we create a class we could get both the Id and the string back by implementing IDataFetch<Genre_DTO> if we had a class called Genre_DTO that had the Id and the Name field (instead of only fetching the Name field)

File: LibraryManagement.Data/DataRequestObjects/Genres/InsertGenres.cs [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace LibraryManagement.Data.DataRequestObjects.Genres
+{
+    public class InsertGenres : IDataExecute
+    {
+        public string _Name { get; set; }
```
Comment (uhyeay2) [inline LibraryManagement.Data/DataRequestObjects/Genres/InsertGenres.cs]: Since this is a public property it would be best to use PascalCase for the naming.... using _cameCase with an underscore is typically for private fields

File: LibraryManagement.Data/DataRequestObjects/Genres/DeleteGenres.cs (hunk 10 above)
Comment (brettmariani923) [reply LibraryManagement.Data/DataRequestObjects/Genres/DeleteGenres.cs:27]: Hey I think the auto incrementing is creating an issue where when I go to delete by ID, the auto incrementing makes the next input resume from the number following the last one that was deleted. For example, if I were to delete the subject in fields 1 and 2, the next input would have a genreID of 3 and 4, and its messing with the ability to delete through my test method.

I'm thinking I should probably just delete the GenreID category, the GenreID is already being covered by the auto incrementing and we have the UNIQUE(NAME) constraint that prevents any duplicate names. I dont see a scenario where we would need multiple categories with the same name so I dont think it should be much of an issue, but if you have a better way to go about it or a reason not to let me know. 

Heres the current code

```CREATE TABLE [dbo].[Genres]
(
    GenreID INT CONSTRAINT PK_Genres PRIMARY KEY IDENTITY(1,1),
    Name NVARCHAR(50) NOT NULL,
    CONSTRAINT CHK_Genres_Name CHECK (LEN(Name) > 0), 
    CONSTRAINT UQ_Genres_Name UNIQUE(Name)
);```

File: LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/InsertGenresTest.cs (hunk 8 above)
Comment (brettmariani923) [reply LibraryManagement.Data.Tests/DataRequestTests/ExampleTests/InsertGenresTest.cs]: Sounds good, I updated to request.Name and added an await here and to the other relevant tests as well.

File: LibraryManagement.Data/DataRequestObjects/Genres/GetGenre.cs (hunk 11 above)
Comment (brettmariani923) [reply LibraryManagement.Data/DataRequestObjects/Genres/GetGenre.cs:10]: I changed the "GetGenres.cs" class to "Genre_DTO.cs", and here is the updated code 
```using LibraryManagement.Data.Abstraction;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LibraryManagement.Data.DataRequestObjects.Genres
{
    public class Genre_DTO : IDataFetch<Genre_DTO>
    {
        public int GenreID { get; set; }

        public Genre_DTO(string genreName, int genreId)
        {
            GenreID = genreId;
        }

        public object GetParameters()
        {
            return new { GenreID = GenreID };
        }

        public string GetSql()
        {
            return "SELECT GenreID, Name FROM Genres WHERE GenreID = @GenreID";
        }

    }
}
```

File: LibraryManagement.Data/DataRequestObjects/Genres/DeleteGenres.cs (hunk 10 above)
Comment (brettmariani923) [reply LibraryManagement.Data/DataRequestObjects/Genres/DeleteGenres.cs:27]: This is the code I'm using 
```namespace LibraryManagement.Data.DataRequestObjects.Genres
{
    public class DeleteGenres : IDataExecute
    {
        public int DeleteID { get; set; }

        public DeleteGenres(int id)
        {
            DeleteID = id;
        }

        public object? GetParameters()
        {
            return new { GenreID = DeleteID };
        }

        public string GetSql()
        {
            return "DELETE FROM Genres WHERE GenreID = @GenreID;";
        }
   
    }
}```

**Review by uhyeay2 — APPROVED** [review]
